hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/KeyPathRetriever.java

```diff
@@ -0,0 +1,209 @@
+import org.apache.hadoop.ozone.om.OmMetadataManagerImpl;
+import org.apache.hadoop.ozone.om.helpers.OmBucketInfo;
+import org.apache.hadoop.ozone.om.helpers.OmDirectoryInfo;
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.kohsuke.MetaInfServices;
+import picocli.CommandLine;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import static org.apache.hadoop.ozone.OzoneConsts.OM_KEY_PREFIX;
+
+/**
+ * Tool that retrieve key full path from OM db file table with pattern match.
+ */
+@CommandLine.Command(
+    name = "retrieve-key-fullpath",
+    description = "retrieve full key path with matching criteria")
+@MetaInfServices(SubcommandWithParent.class)
+public class KeyPathRetriever implements Callable<Void>, SubcommandWithParent {
+  @CommandLine.Spec
+  private static CommandLine.Model.CommandSpec spec;
+
+  @CommandLine.Option(names = {"--db"},
+      required = true,
+      description = "Database File")
+  private String dbFile;
+
+  @CommandLine.Option(names = {"--containers"},
+      required = false,
+      description = "Comma separated Container Ids")
+  private String strContainerIds;
+
+  @CommandLine.Option(names = {"--container-file"},
+      required = false,
+      description = "file with container id in new line")
+  private String containerFileName;
+
+  @CommandLine.Option(names = {"--out", "-o"},
+      required = false,
+      description = "out file name")
+  private String fileName;
+
+  @Override
+  public Class<?> getParentType() {
+    return OzoneDebug.class;
+  }
+
+  @Override
+  public Void call() throws Exception {
+    // get containerIds filter
+    Set<Long> containerIds = new HashSet<>();
+    if (!StringUtils.isEmpty(strContainerIds)) {
+      if (!StringUtils.isEmpty(strContainerIds)) {
+        String[] split = strContainerIds.split(",");
+        for (String id : split) {
+          containerIds.add(Long.parseLong(id));
+        }
+      }
+    } else if (!StringUtils.isEmpty(containerFileName)) {
+      try (Stream<String> stream = Files.lines(Paths.get(containerFileName))) {
+        stream.forEach(e -> containerIds.add(Long.parseLong(e)));
+      }
+    }
+    if (containerIds.isEmpty()) {
+      System.out.println("No containers to be filtered");
+      return null;
+    }
+    // out stream
+    PrintWriter writer;
+    if (StringUtils.isEmpty(fileName)) {
+      writer = out();
+    } else {
+      writer = new PrintWriter(new BufferedWriter(new PrintWriter(fileName, UTF_8.name())));
+    }
+
+    // db handler
+    OMMetadataManager metadataManager = getOmMetadataManager(dbFile);
+    if (metadataManager == null) {
+      writer.close();
+      return null;
+    }
+
+    try {
+      retrieve(metadataManager, writer, containerIds);
+    } finally {
+      writer.close();
+      metadataManager.stop();
+    }
+    return null;
+  }
+
+  public void retrieve(
+      OMMetadataManager metadataManager, PrintWriter writer, Set<Long> containerIds) {
```

Comment:
I'm ok with this PR having its own iteration logic that may not be optimal in order to get a working version out. However, in a follow up PR we should try to put the iteration logic for FSO and OBS buckets in a common place so other tools like this and fso-repair can use it.

The shared implementation could take a consumer to process each key, a thread count, and optional parameters could limit it by volume or bucket. IMO speed and low memory overhead are more important than sorted order in these cases, so we can parallelize the tree search if we drop these requirement.

Reply:
@errose28 added JIRA to track